File: Projects/Week 6 - Python/sentimental-credit/credit.py
from cs50 import get_string

# Get credit card number from user
number = get_string("Number: ")

# Function to validate card number using Luhn's algorithm


def luhn_checksum(card_number):
    total = 0
    reverse_digits = card_number[::-1]  # Reverse the card number for easy indexing

    for i, digit in enumerate(reverse_digits):
        n = int(digit)

        # Double every second digit starting from the right
        if i % 2 == 1:
            n *= 2
            if n > 9:
                n -= 9

        total += n

    # Card is valid if total is a multiple of 10
    return total % 10 == 0


# Check if the card number is valid
if luhn_checksum(number):
    # Check for AMEX
    if len(number) == 15 and (number.startswith("34") or number.startswith("37")):
        print("AMEX")
    # Check for MASTERCARD
    elif len(number) == 16 and number.startswith(("51", "52", "53", "54", "55")):
        print("MASTERCARD")
    # Check for VISA
    elif len(number) in [13, 16] and number.startswith("4"):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: Projects/Week 1 - C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long long card_number;
    int sum = 0;
    int length = 0;

    // Prompt the user for the credit card number
    do
    {
        card_number = get_long_long("Number: ");
    }
    while (card_number <= 0);

    long long temp_card = card_number;

    // Check the length of the card number and calculate checksum
    while (temp_card > 0)
    {
        length++;
        int digit = temp_card % 10; // Get the last digit

        // If the position of the digit is even (starting from the end)
        if ((length % 2) == 0)
        {
            digit *= 2; // Double the digit
            // If the doubled digit is greater than 9, add the digits
            sum += (digit > 9) ? (digit - 9) : digit;
        }
        else
        {
            sum += digit; // Add the digit directly
        }
        temp_card /= 10; // Remove the last digit
    }

    // Validate the card number
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Determine the card type
    if ((length == 15) &&
        (card_number / 10000000000000 == 34 || card_number / 10000000000000 == 37))
    {
        printf("AMEX\n");
    }
    else if ((length == 16) &&
             (card_number / 100000000000000 >= 51 && card_number / 100000000000000 <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) &&
             (card_number / 1000000000000 == 4 || card_number / 1000000000000000 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
